ext/ScreenShot/ScreenShot.h

```diff
@@ -5,72 +5,186 @@
+			core::smart_refctd_ptr<video::IDriverFence> downloadImageMipLevel(video::IDriver* driver, video::IGPUImage* source, video::IGPUBuffer* destination, uint32_t sourceMipLevel = 0u, size_t destOffset = 0ull, bool implicitflush = true)
+			{
+				// will change this, https://github.com/buildaworldnet/IrrlichtBAW/issues/148
+				if (isBlockCompressionFormat(source->getCreationParameters().format))
+					return nullptr;
+
+				auto extent = source->getMipSize(sourceMipLevel);
+				video::IGPUImage::SBufferCopy pRegions[1u] = { {destOffset,extent.x,extent.y,{static_cast<asset::IImage::E_ASPECT_FLAGS>(0u),sourceMipLevel,0u,1u},{0u,0u,0u},{extent.x,extent.y,extent.z}} };
+				driver->copyImageToBuffer(source, destination, 1u, pRegions);
+
+				return driver->placeFence(implicitflush);
+			}
+
+			/*
+				Create a ScreenShot with gpu image usage and save it to a file.
+			*/
+
+			bool createScreenShot(core::smart_refctd_ptr<IrrlichtDevice> device, const video::IGPUImageView* gpuImageView, const std::string& outFileName)
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageViewParmas = gpuImageView->getCreationParameters();
+				auto gpuImage = fetchedImageViewParmas.image;
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				video::IDriverMemoryBacked::SDriverMemoryRequirements memoryRequirements;
+				memoryRequirements.vulkanReqs.alignment = 64u;
+				memoryRequirements.vulkanReqs.memoryTypeBits = 0xffffffffu;
+				memoryRequirements.memoryHeapLocation = video::IDriverMemoryAllocation::ESMT_NOT_DEVICE_LOCAL;
+				memoryRequirements.mappingCapability = video::IDriverMemoryAllocation::EMCF_CAN_MAP_FOR_READ | video::IDriverMemoryAllocation::EMCF_COHERENT | video::IDriverMemoryAllocation::EMCF_CACHED;
+				memoryRequirements.vulkanReqs.size = image->getImageDataSizeInBytes();
+				auto destinationBuffer = driver->createGPUBufferOnDedMem(memoryRequirements);
+
+				auto fence = downloadImageMipLevel(driver, gpuImage.get(), destinationBuffer.get());
+				while (fence->waitCPU(1000ull, fence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
```

Comment:
you don't need the fence here, the only thing you want to wait on is accessing the `texelBuffer->getPointer()` pointer because you don't want to look at (read) that memory area before the GPU writes to it

all the mapping, etc. can still be going on while GPU is working.